### ISSUE TITLE
[Question] Is Linux ARM64 a supported architecture?

### ISSUE THREAD

**Issue by markjens** [afni/afni#390]
Hello,

I’d like to ask whether afni is supported on Linux/Mac ARM64 systems ?
I see `src/Makefiles*` for Mac M1 any many others, but none of them look appropriate for Linux ARM64.

Thank you!

**Comment by afni-rickr** [comment]
We have no such device at the moment, but could look into it.  What laptop/desktop/other device(s) are you picturing running it on?

**Comment by neurolabusc** [comment]
@markjens AFNI does compile for MacOS ARM64 systems. Simply use the [Makefile.macos_11_ARM_clang](https://github.com/afni/afni/blob/master/src/Makefile.macos_11_ARM_clang). The Apple [M1 has outstanding floating point performance](https://github.com/neurolabusc/AppleSiliconForNeuroimaging). 

I have not tried compiling AFNI for a Linux ARM64. I suspect this will require a new LINUX makefile that uses clang. Most non-Apple inexpensive ARM computers (like the [Pinebook](https://www.pine64.org/pinebook-pro/) use CPUs that do not have great floating point performance. However, I think that the Oracle and Amazon cloud ARM CPUs should be very capable. 

One option to evaluate this would be to use the in-built virtualization of the latest MacOS to run Linux. For example, [this video](https://developer.apple.com/videos/play/wwdc2022/10002/) details how you can run Linux as a guest on a M1 running MacOS, with an implementation [here](https://developer.apple.com/documentation/virtualization/running_gui_linux_in_a_virtual_machine_on_a_mac). This would allow you to test the ARM Linux build scripts before you deploy them onto the cloud.

**Comment by neurolabusc** [comment]
@markjens yes it is possible, with the (solveable) caveats listed below.

@afni-rickr and @afni-dglen perhaps we can work together to clean up my rough solution. Oracle is graciously providing development time on their [80-core 3 GHz Ampere Altra A1](https://www.oracle.com/cloud/compute/arm/) systems (with bare metal allocation allowing a 2-socket 160 core and 1TB of RAM). You can request a server via [brainhack_cloud](https://github.com/brainhackorg/brainhack_cloud/issues). I am happy to walk you through the setup. You can select your desired Linux image including Debian-based Ubuntu or Fedora-based CentOS or Oracle. For my testing I used Oracle 8.6.

There are several caveats:

 - Unable to find these packages `sudo yum install PyQt4 R-devel`, though these may be more used in GUI situations rather than on cloud clusters.
 - Need to remove `-DHAVE_XDBE` from makefile, otherwise "fatal error: 'X11/extensions/Xdbe.h' file not found"
 - In makefile, add line `CPU_TYPE = ARM_M1`. I realize Apple M1 is different from Altra, so maybe AFNI should define a general ARM64 type.
 - gcc 8.5.0 for ARM does not support the [x86 Option](https://gcc.gnu.org/onlinedocs/gcc/x86-Options.html) `-m64` directive so I removed it from the gcc makefile.
 - clang 13.0 does not complain about `-m64` not happy with `omp.h`. Therefore, my recipe is for gcc. I could not work out how to get OpenMP to work on this arm64 system. We could explore this together (perhaps `sudo yum install clang llvm-toolset`)
 - Both clang and gcc (without `-m64`) compile many libraries. But both eventually fail with the [same error](https://github.com/afni/afni/blob/7a5da50e4f7968e0bf526add8ad932da67cab4ad/src/suma_utils.c#L3048): `suma_utils.c:3048:14: error: used struct type value where scalar is required`. Since this is just a conditional to choose between error messages, I removed the comparison and made an error message that said `SUMA_LH("vararg_ptr exception");`
 - I get a linker error `/usr/bin/ld: cannot find -lpython2.7` which seems to be fixed with `yum install python2-devel`, however python2.7 is EOL, so perhaps this should be fixed at the source.
 - Attempting to compile AFNI in parallel using `make vastness -j`fails. Does anyone know if it is possible to compile the AFNI libraries and dependencies sequentially and then all the AFNI executables in in parallel? This would save a lot of time on modern computers.
 - After compiling, one gets the warning `checking for GLIB - version >= 2.4.0... no` however Package glib2-2.56.4-158.el8.aarch64 is already installed.
 - After compilation there is a warning `configure: WARNING: libnetpbm`` but netpbm-10.82.00-7.el8.aarch64 is already installed.
 - After compiling, there are warnings that libgts.a is missing.
 - In general, virtually all executables are created and 3dDeconvolve seems to work as expected.
 
Experimental makefile:

```
# This Makefile was written for the Linux OS, fedora-like ARM64 Oracle Linux Server 8.6.
# It was originally based off linux_openmp_64.

SYSTEM_NAME = linux_ARM_64
SYSTEM_TYPE = Linux

# CPROF = -pg -g
USE_ZLIB = -DHAVE_ZLIB
LZLIB    = -lz
USE_GIFTI = -DHAVE_GIFTI
LGIFTI    = -lexpat


# -----------------------------------
#  # use shared libriaries (libmri.so)?
MRI_SHARED = $(SYSTEM_TYPE)

# use linux-style R_io.so building
R_IO_MODIFY_LINUX = yes

# ------------------------------
# python from C
IPYTHON  = -DSELENIUM_READY -I/usr/include/python2.7
LDPYTHON = -lpython2.7

# ----------------------------------------------------------------------
# X configuration
XLIBS = -lXm -lXt

# ----------------------------------------------------------------------

CCDEBS = -DAFNI_DEBUG -DIMSEQ_DEBUG -DDISPLAY_DEBUG -DTHD_DEBUG
CEXTRA = -Wcomment -Wformat -DUSE_TRACING $(CPROF) \
         -DDONT_USE_XTDESTROY -D_GNU_SOURCE -DREPLACE_XT

CC     = /usr/bin/gcc -O2 -fPIC -DREAD_WRITE_64 -DLINUX2 $(CEXTRA)
CCVOL  = $(CC)
CCFAST = $(CC)
CCOLD  = $(CC)

# set CPU_TYPE for libgts.a building and anything else that might need it
CPU_TYPE = ARM_M1

# The following line includes compiling for the SSE operations.
# However, I found that it actually makes things worse in some test code.
# Your mileage may vary.

OMPFLAG = -fopenmp -DUSE_OMP
CCMIN  = /usr/bin/gcc  -fPIC $(CPROF)
CCD    = $(CC) $(CCDEBS)

IFLAGS = -I. -I/usr/include
LFLAGS = -L. -L/usr/lib64

CCSVD  = /usr/bin/gcc -O0 -fPIC

PLUGIN_SUFFIX = so
PLUGIN_LFLAGS = -shared -fPIC
PLUGIN_CC     = $(CC)
PLFLAGS       = -rdynamic -L. -L/usr/lib64

# possibly modify from above
ifdef MRI_SHARED
LFLAGS  = -L. -L/usr/lib64 -Wl,-rpath='$$ORIGIN'
PLFLAGS = -rdynamic $(LFLAGS)
endif

SHOWOFF = -DSHOWOFF=$(SYSTEM_NAME)

AR     = /usr/bin/ar
RANLIB = /usr/bin/ranlib
TAR    = /bin/tar
MKDIR  = /bin/mkdir
GZIP   = /bin/gzip
LD     = /usr/bin/gcc $(CPROF)

RM = /bin/rm -f
MV = /bin/mv -f
CP = /bin/cp -f

LINT = /usr/bin/lint -a -b -u -v -x $(IFLAGS) $(CCDEFS)

INSTALLDIR = ./$(SYSTEM_NAME)
LIBDIR = $(INSTALLDIR)

#INSTALL_PREREQ = suma
INSTALL_PREREQ = suma_gts
# uncomment if the Gnu Scientific Library is installed (libgsl, libgslcblas)
GSLPROGS = balloon
EXPROGS = $(GSLPROGS)

# for pure dynamic linking

# LLIBS  = -lmri -lf2c -lXm -lXft -lXp -lXpm -lXext -lXmu -lXt -lSM -lICE -lX11 -lpng -ljpeg $(LZLIB) $(LGIFTI) -lm  -ldl -lc

# link choices are made above

# separate X11 libs from non-X11 ones
LLIBS  = -lmri -lf2c $(LDPYTHON) -lpng -ljpeg $(LZLIB) $(LGIFTI) -lm  -ldl -lc
LLIBS_X11 = -lmri -lf2c $(XLIBS) -lXft -lXpm -lXext -lXmu -lSM -lICE -lX11 \
         $(LDPYTHON) -lpng -ljpeg $(LZLIB) $(LGIFTI) -lm  -ldl -lc

# for static linking, as far as possible

# LLIBS = -lmri -lf2c -ldl               \
#        /usr/lib/libXm.a   \
#        /usr/lib/libXpm.a  \
#        /usr/lib/libXext.a \
#        /usr/lib/libXmu.a  \
#        /usr/lib/libXt.a   \
#        /usr/lib/libSM.a   \
#        /usr/lib/libICE.a  \
#        /usr/lib/libX11.a  \
#        /usr/lib/libm.a          \
#        /usr/lib/libc.a

# vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv
# For suma (NO STATIC LINKING OF GL libs)
SUMA_GLIB_VER = -2.0

#use -lGLw if you have libGLw.a or libGLw.so* or 
#  -lMesaGLw if you have Mesa's version (libMesaGLw*) of libGLw
# GLw_IPATH  = -IGLw_local
GLw_LIB = -lGLw

SUMA_INCLUDE_PATH = $(IFLAGS) -I.. -I../niml -Igts/src 		\
    -I/usr/lib64/glib-2.0/include -I/usr/lib64/glib/include	\
    -I/usr/include/glib-2.0 -I/usr/include/glib-1.2
SUMA_LINK_PATH = -L.. $(LFLAGS)
#use -lGLw if you have libGLw.a or libGLw.so* or 
#  -lMesaGLw if you have Mesa's version (libMesaGLw*) of libGLw
SUMA_LINK_LIB = $(GLw_LIB) -lGLU -lGL $(LLIBS_X11)
SUMA_MAKEFILE_NAME = SUMA_Makefile
SUMA_BIN_ARCHIVE = SUMA_Linux.tar
SUMA_MDEFS = 
# ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

###############################################################

MAKE = make
include Makefile.INCLUDE
```

**Comment by neurolabusc** [comment]
Oracle also provides a Debian-based Ubuntu 22.04 image with gcc 11.2. I focused on getting `make all` work: I think I will need help for `make vastness`.

 - I started with the [AFNI Ubuntu 20.04 instructions](https://afni.nimh.nih.gov/pub/dist/doc/htmldoc/background_install/install_instructs/steps_linux_ubuntu20.html)
 - no tweak tool, no fortran
 - again, need to remove `-m64` compiler flag from makefile
 - gcc (without -m64) compiles the tools. But eventually fails at suma_utils.c:3048:14: error: used struct type value where scalar is required. Since this is just a conditional to choose between error messages, I removed the comparison and made an error message that said SUMA_LH("vararg_ptr exception");
 -  `make vastness` complains about `multiple definition of `glwDrawingAreaWidgetClass'` in many objects (SUMA_Color.o, SUMA_IsoSurfaceFunc.o, SUMA_VolData.o, etc, etc.
 
**makefile:**

```
# for Ubuntu 22, based on other_builds/Makefile.linux_ubuntu_12_64_OMP

SYSTEM_NAME = linux_ubuntu_ARM64
SYSTEM_TYPE = Linux

# CPROF = -pg -g
USE_ZLIB = -DHAVE_ZLIB
LZLIB    = -lz
USE_GIFTI = -DHAVE_GIFTI
LGIFTI    = -lexpat  #/usr/lib64/libexpat.a

# -----------------------------------
#  # use shared libriaries (libmri.so)?
MRI_SHARED = $(SYSTEM_TYPE)

# use linux-style R_io.so building
R_IO_MODIFY_LINUX = yes

# -----------------------------------
# default to having these off
# D_USE_ACML = -DUSE_ACML
# I_USE_ACML = -I/opt/acml3.1.0/gnu64/include
# L_USE_ACML = /opt/acml3.1.0/gnu64/lib/libacml.a

CCDEBS = -DAFNI_DEBUG -DIMSEQ_DEBUG -DDISPLAY_DEBUG -DTHD_DEBUG
CEXTRA = -Wcomment -Wformat -DUSE_TRACING -DHAVE_XDBE -D_GNU_SOURCE \
         -DREPLACE_XT $(CPROF)

CC     = /usr/bin/gcc -O2 -fPIC -DREAD_WRITE_64 -DLINUX2 $(CEXTRA)
CCVOL  = /usr/bin/gcc -O2 -fPIC -DREAD_WRITE_64 -DLINUX2 $(CEXTRA)
CCFAST = /usr/bin/gcc -O2 -fPIC -DREAD_WRITE_64 -DLINUX2 $(CEXTRA)
CCOLD  = $(CC)


# The following line includes compiling for the SSE operations.
# However, I found that it actually makes things worse in some test code.
# Your mileage may vary.

# CCFAST = /usr/bin/gcc -O3 -march=i686 -ffast-math -mmmx -msse -mfpmath=sse -DLINUX2 $(CEXTRA)

OMPFLAG = -fopenmp -DUSE_OMP  # for OpenMP
CCMIN  = /usr/bin/gcc -fPIC $(CPROF)
CCD    = $(CC) $(CCDEBS)
IFLAGS = -I. -I/usr/X11R6/include $(I_USE_ACML)
#-I/usr/src/linux-headers-3.0.0-17/include/linux -I/usr/src/linux-headers-3.0.0-17/include $(I_USE_ACML)
LFLAGS = -L. -L/usr/X11R6/lib64 $(L_USE_ACML)

# maybe 

CCSVD  = /usr/bin/gcc -O0 -fPIC

PLUGIN_SUFFIX = so
PLUGIN_LFLAGS = -shared -fPIC
PLUGIN_CC     = $(CC)
# PLFLAGS       = -rdynamic -L. -L/usr/X11R6/lib
PLFLAGS       = -rdynamic -L. -L/usr/X11R6/lib64

# possibly modify from above
ifdef MRI_SHARED
LFLAGS  = -L. -L/usr/lib64 $(L_USE_ACML) -Wl,-rpath='$$ORIGIN'
PLFLAGS = -rdynamic $(LFLAGS)
endif

SHOWOFF = -DSHOWOFF=$(SYSTEM_NAME)

AR     = /usr/bin/ar
RANLIB = /usr/bin/ranlib
TAR    = /bin/tar
MKDIR  = /bin/mkdir
GZIP   = /bin/gzip
LD     = /usr/bin/gcc $(CPROF)

RM = /bin/rm -f
MV = /bin/mv -f
CP = /bin/cp -f

LINT = /usr/bin/lint -a -b -u -v -x $(IFLAGS) $(CCDEFS)

INSTALLDIR = ./$(SYSTEM_NAME)
LIBDIR = $(INSTALLDIR)

# INSTALL_PREREQ = suma
INSTALL_PREREQ = suma_gts
# uncomment if the Gnu Scientific Library is installed (libgsl, libgslcblas)
 GSLPROGS = balloon
EXPROGS = $(GSLPROGS)   # 3dQwarp gifti_tool $(GSLPROGS)

# for dynamic linking

# LLIBS  = -lmri -lf2c -lXm -lXpm -lXext -lXmu -lXt -lX11 $(LZLIB) $(LGIFTI) -lm  -ldl -lc

# for static linking to Motif, dynamic to all else

# separate X11 libs from non-X11 ones
LLIBS  = -lmri -lf2c $(LZLIB) $(LGIFTI) -lm  -ldl -lc
LLIBS_X11 = -lmri -lf2c -lXm -lXpm -lXext -lXmu -lXt -lSM -lICE -lX11 \
            $(LZLIB) $(LGIFTI) -lm  -ldl -lc

#LLIBS  = -lmri -lf2c /usr/X11R6/lib64/libXm.a -lXpm -lXext -lXmu -lXt -lSM -lICE -lX11 $(LZLIB) $(LGIFTI) -lm  -ldl -lc

# for static linking, as far as possible

# LLIBS = -lmri -lf2c -ldl               \
#         /usr/X11R6/lib/libXm.a   \
#        /usr/X11R6/lib/libXpm.a  \
#        /usr/X11R6/lib/libXext.a \
#        /usr/X11R6/lib/libXmu.a  \
#        /usr/X11R6/lib/libXt.a   \
#        /usr/X11R6/lib/libSM.a   \
#        /usr/X11R6/lib/libICE.a  \
#        /usr/X11R6/lib/libX11.a  \
#        /usr/lib/libm.a          \
#        /usr/lib/libc.a

# vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv
# For suma (NO STATIC LINKING OF GL libs)
SUMA_GLIB_VER = -2.0
SUMA_INCLUDE_PATH = -I/usr/X11R6/include -I./ -I../ -I../niml/ -Igts/src -I/usr/include/glib-1.2 -I/usr/include/glib-2.0 -I/usr/l
ib64/glib/include -I/usr/lib64/glib-2.0/include -I/usr/lib/x86_64-linux-gnu/glib-2.0/include/
SUMA_LINK_PATH = -L/usr/lib64 -L/usr/X11R6/lib64 -L../
#use -lGLw if you have libGLw.a or libGLw.so* or 
#  -lMesaGLw if you have Mesa's version (libMesaGLw*) of libGLw
#
SUMA_LINK_LIB = -lGLw -lGLU -lGL $(LLIBS_X11)
#SUMA_LINK_LIB = /usr/X11R6/lib64/libXm.a -lGLw -lGLU -lGL -lXmu -lXt -lXext -lX11 $(LZLIB) $(LGIFTI) -lm 
SUMA_MAKEFILE_NAME = SUMA_Makefile
SUMA_BIN_ARCHIVE = SUMA_Linux.tar
SUMA_MDEFS = 
# ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

###############################################################

MAKE = make
include Makefile.INCLUDE
```

**recipe:**

```
sudo apt-get update
sudo apt-get install -y tcsh xfonts-base libssl-dev       \
                        python-is-python3                 \
                        python3-matplotlib python3-numpy  \
                        gsl-bin netpbm   \
                        libjpeg62 xvfb xterm vim curl     \
                        gedit evince eog                  \
                        libglu1-mesa-dev libglw1-mesa     \
                        libxm4 build-essential            \
                        libcurl4-openssl-dev libxml2-dev  \
                        libgomp1        \
                        gnome-terminal nautilus           \
                        gnome-icon-theme-symbolic         \
                        firefox xfonts-100dpi             \
                        r-base-dev cmake                  \
                        libgdal-dev libopenblas-dev       \
                        libnode-dev libudunits2-dev
sudo apt-get install -y libxext-dev
sudo apt install -y libmotif-dev
sudo apt install -y libxpm-dev
sudo apt install -y libxmu-dev libxmu-headers
sudo apt -y install libgsl-dev

git clone https://github.com/afni/afni.git
cp Makefile.linux_ubuntu_22_ARM64 Makefile
#at this stage you can "make all"
sudo apt-get install libxi-dev
sudo apt -y install libglw1-mesa-dev
sudo apt install libx11-dev
make vastness
```

**Comment by neurolabusc** [comment]
I was also able to get AFNI to compile with Ubuntu using `clang` instead of `gcc`.  The recipe is the same as before with the addition of:

```
sudo apt-get install clang libomp-dev libomp-dev
```

The makefile is again derived from the existing Ubuntu makefile, but one can retain the `-m64` flag and change all instances of `gcc` to `clang`:

```
# for Ubuntu 22 using clang, based on other_builds/Makefile.linux_ubuntu_12_64_OMP

SYSTEM_NAME = linux_ubuntu_16_64
SYSTEM_TYPE = Linux

# CPROF = -pg -g
USE_ZLIB = -DHAVE_ZLIB
LZLIB    = -lz
USE_GIFTI = -DHAVE_GIFTI
LGIFTI    = -lexpat  #/usr/lib64/libexpat.a

# -----------------------------------
#  # use shared libriaries (libmri.so)?
MRI_SHARED = $(SYSTEM_TYPE)

# use linux-style R_io.so building
R_IO_MODIFY_LINUX = yes

# -----------------------------------
# default to having these off
# D_USE_ACML = -DUSE_ACML
# I_USE_ACML = -I/opt/acml3.1.0/gnu64/include
# L_USE_ACML = /opt/acml3.1.0/gnu64/lib/libacml.a

CCDEBS = -DAFNI_DEBUG -DIMSEQ_DEBUG -DDISPLAY_DEBUG -DTHD_DEBUG
CEXTRA = -Wcomment -Wformat -DUSE_TRACING -DHAVE_XDBE -D_GNU_SOURCE \
         -DREPLACE_XT $(CPROF)

CC     = /usr/bin/clang -O2 -m64 -fPIC -DREAD_WRITE_64 -DLINUX2 $(CEXTRA)
CCVOL  = /usr/bin/clang -O2 -m64 -fPIC -DREAD_WRITE_64 -DLINUX2 $(CEXTRA)
CCFAST = /usr/bin/clang -O2 -m64 -fPIC -DREAD_WRITE_64 -DLINUX2 $(CEXTRA)
CCOLD  = $(CC)


# The following line includes compiling for the SSE operations.
# However, I found that it actually makes things worse in some test code.
# Your mileage may vary.

# CCFAST = /usr/bin/clang -O3 -march=i686 -ffast-math -mmmx -msse -mfpmath=sse -
DLINUX2 $(CEXTRA)

OMPFLAG = -fopenmp -DUSE_OMP  # for OpenMP
CCMIN  = /usr/bin/clang -m64 -fPIC $(CPROF)
CCD    = $(CC) $(CCDEBS)
IFLAGS = -I. -I/usr/X11R6/include $(I_USE_ACML)
#-I/usr/src/linux-headers-3.0.0-17/include/linux -I/usr/src/linux-headers-3.0.0-
17/include $(I_USE_ACML)
LFLAGS = -L. -L/usr/X11R6/lib64 $(L_USE_ACML)

# maybe 

CCSVD  = /usr/bin/clang -O0 -m64 -fPIC

PLUGIN_SUFFIX = so
PLUGIN_LFLAGS = -shared -fPIC
PLUGIN_CC     = $(CC)
# PLFLAGS       = -rdynamic -L. -L/usr/X11R6/lib
PLFLAGS       = -rdynamic -L. -L/usr/X11R6/lib64

# possibly modify from above
ifdef MRI_SHARED
LFLAGS  = -L. -L/usr/lib64 $(L_USE_ACML) -Wl,-rpath='$$ORIGIN'
PLFLAGS = -rdynamic $(LFLAGS)
endif

SHOWOFF = -DSHOWOFF=$(SYSTEM_NAME)

AR     = /usr/bin/ar
RANLIB = /usr/bin/ranlib
TAR    = /bin/tar
MKDIR  = /bin/mkdir
GZIP   = /bin/gzip
LD     = /usr/bin/clang $(CPROF)

RM = /bin/rm -f
MV = /bin/mv -f
CP = /bin/cp -f

LINT = /usr/bin/lint -a -b -u -v -x $(IFLAGS) $(CCDEFS)

INSTALLDIR = ./$(SYSTEM_NAME)
LIBDIR = $(INSTALLDIR)

# INSTALL_PREREQ = suma
INSTALL_PREREQ = suma_gts
# uncomment if the Gnu Scientific Library is installed (libgsl, libgslcblas)
 GSLPROGS = balloon
EXPROGS = $(GSLPROGS)   # 3dQwarp gifti_tool $(GSLPROGS)

# for dynamic linking

# LLIBS  = -lmri -lf2c -lXm -lXpm -lXext -lXmu -lXt -lX11 $(LZLIB) $(LGIFTI) -lm
  -ldl -lc

# for static linking to Motif, dynamic to all else

# separate X11 libs from non-X11 ones
LLIBS  = -lmri -lf2c $(LZLIB) $(LGIFTI) -lm  -ldl -lc
LLIBS_X11 = -lmri -lf2c -lXm -lXpm -lXext -lXmu -lXt -lSM -lICE -lX11 \
            $(LZLIB) $(LGIFTI) -lm  -ldl -lc

#LLIBS  = -lmri -lf2c /usr/X11R6/lib64/libXm.a -lXpm -lXext -lXmu -lXt -lSM -lIC
E -lX11 $(LZLIB) $(LGIFTI) -lm  -ldl -lc

# for static linking, as far as possible

# LLIBS = -lmri -lf2c -ldl               \
#         /usr/X11R6/lib/libXm.a   \
#        /usr/X11R6/lib/libXpm.a  \
#        /usr/X11R6/lib/libXext.a \
#        /usr/X11R6/lib/libXmu.a  \
#        /usr/X11R6/lib/libXt.a   \
#        /usr/X11R6/lib/libSM.a   \
#        /usr/X11R6/lib/libICE.a  \
#        /usr/X11R6/lib/libX11.a  \
#        /usr/lib/libm.a          \
#        /usr/lib/libc.a

# vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv
# For suma (NO STATIC LINKING OF GL libs)
SUMA_GLIB_VER = -2.0
SUMA_INCLUDE_PATH = -I/usr/X11R6/include -I./ -I../ -I../niml/ -Igts/src -I/usr/
include/glib-1.2 -I/usr/include/glib-2.0 -I/usr/lib64/glib/include -I/usr/lib64/
glib-2.0/include -I/usr/lib/x86_64-linux-gnu/glib-2.0/include/
SUMA_LINK_PATH = -L/usr/lib64 -L/usr/X11R6/lib64 -L../
#use -lGLw if you have libGLw.a or libGLw.so* or 
#  -lMesaGLw if you have Mesa's version (libMesaGLw*) of libGLw
#
SUMA_LINK_LIB = -lGLw -lGLU -lGL $(LLIBS_X11)
#SUMA_LINK_LIB = /usr/X11R6/lib64/libXm.a -lGLw -lGLU -lGL -lXmu -lXt -lXext -lX
11 $(LZLIB) $(LGIFTI) -lm 
SUMA_MAKEFILE_NAME = SUMA_Makefile
SUMA_BIN_ARCHIVE = SUMA_Linux.tar
SUMA_MDEFS = 
# ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

###############################################################

MAKE = make
include Makefile.INCLUDE
```

I tested an Ampere Altra instance with clang 14.0 and gcc 11.0. They performed very similarly on the AFNI 3dDeconvolve benchmark. I summarize [overall benchmarks of the Ampere Altra here](https://github.com/neurolabusc/AppleSiliconForNeuroimaging).

**Comment by martin-g** [comment]
> Oracle is graciously providing development time on their [80-core 3 GHz Ampere Altra A1](https://www.oracle.com/cloud/compute/arm/) systems (with bare metal allocation allowing a 2-socket 160 core and 1TB of RAM). You can request a server via [brainhack_cloud](https://github.com/brainhackorg/brainhack_cloud/issues).

Hi!
Oracle Cloud provides free Linux ARM64 VMs - https://martin-grigorov.medium.com/github-actions-arm64-runner-on-oracle-cloud-a77cdf7a325a.
The tiny problem is that they require credit card details to create an account (as any other cloud provider does)! But as long as you don't use paid resources they do not charge the credit card.
I am **not** related to Oracle anyhow except that I contribute to OpenJDK project by testing several big Java projects on latest OpenJDK aarch64 builds.

But my employer (Huawei Cloud) is willing to donate free VM(s) that could be used for development needs (testing, debugging, CI, ...) long-term.
If the project team is interested just contact me here or by email (mgrigorov at apache dot org)!

**Comment by martin-g** [comment]
I also see that you use [CircleCI ](https://app.circleci.com/pipelines/github/afni/afni)!
That's great because they provide [ARM64 containers](https://circleci.com/docs/executor-intro#arm)!
One can SSH to a runner VM and debug around!

**Comment by afni-rickr** [comment]
It looked like @neurolabusc wanted that Makefile put in, so it was put in as Makefile.linux_ubuntu_22_ARM_clang, if that seems reasonable, committed on Friday.

**Comment by neurolabusc** [comment]
@afni-rickr 
 1. Github issues will insert add line endings. In this case, `SUMA_INCLUDE_PATH` got corrupted. Can you use the attached text file.
 2. This new Makefile has been tested and will work with either gcc or clang, and with either x86-64 or ARM aarch64, it has been tested with Ubuntu 18.04 through 22.04.
 4. The Makefile will use either clang or gcc, and default to the latter unless the user demands clang (e.g. `useClang=1 make all`).
 5. The error with suma_utils.c described above still exists, and should be fixed in the AFNI repository.
 6. I have removed the `-m64` directive as this was used to coerce 32-bit x86 machines to generate x86-64 code. In modern times, I think it is safe to assume that anyone compiling AFNI will be using a 64-bit machine for both compiling and running. This allows a universal Makefile that works with clang and gcc. 

**Recipe**

```
git clone https://github.com/afni/afni.git
cd afni/src
cp Makefile.linux_ubuntu_22 Makefile
sudo apt-get update
sudo apt-get install -y tcsh xfonts-base libssl-dev       \
                        python3-matplotlib python3-numpy  \
                        gsl-bin netpbm   \
                        libjpeg62 xvfb xterm vim curl     \
                        gedit evince eog                  \
                        libglu1-mesa-dev libglw1-mesa     \
                        libxm4 build-essential            \
                        libcurl4-openssl-dev libxml2-dev  \
                        gnome-terminal nautilus libgomp1  \
                        gnome-icon-theme-symbolic         \
                        firefox xfonts-100dpi libxtst-dev \
                        r-base-dev cmake libxt-dev        \
                        libgdal-dev libopenblas-dev       \
                        libudunits2-dev       \
                        libxext-dev libmotif-dev          \
                        libxpm-dev libxmu-dev libz-dev    \
                        libxmu-headers libgsl-dev make    \
                        libexpat1-dev libmotif-dev
#for Ubuntu 22.04
sudo apt-get install -y python-is-python3 libnode-dev 
#for gcc build
make all
#for clang build
sudo apt-get install -y clang libomp-dev libomp-dev
useClang=1 make all
```

[Makefile.linux_ubuntu_22.txt](https://github.com/afni/afni/files/9350986/Makefile.linux_ubuntu_22.txt)

**Comment by afni-rickr** [comment]
Sorry @markjens. Thanks for the updates, @neurolabusc, they are in.  I should have done it as a separate pull request, but too late now.  @markjens if you want to give Makefile.linux_ubuntu_22_64 a try, please let us know how it goes.  Thanks!

**Comment by neurolabusc** [comment]
@markjens that particular dependency requires either the motif or lesstif headers, which I installed on Debian from `libmotif-dev` and for Red-hat distributions with `motif-devel`. 

I would strongly suggest you install all the libraries in my recipes above: even those that are not explicitly required for compilation may be required dynamically for using AFNI's Python, R and chauffeur_afni. For Red-hat I used a stock Oracle cloud instance to have a clean slate, and I did the same with Oracle cloud instances for Ubuntu 18.04 and 22.04. 

Maybe @afni-rickr can comment about whether evince, gedit, etc are really needed. I included them because they are explicitly listed in the [AFNI 16.04 recipe](https://afni.nimh.nih.gov/pub/dist/doc/htmldoc/background_install/install_instructs/steps_linux_ubuntu16.html) which makes me think they may be required dynamically. My approach was to begin with an absolutely minimal cloud instance, add the dependencies listed by the AFNI recipes, and then only add the dependencies required to compile the code.

In my experience, complex tools like AFNI and FSL do rely on a lot of libraries dynamically, so I would heed their recommendation. We also need to be cautious as recent distributions start dropping tools that used to be installed by default. For example, FSL's Sienna requires the command line `dc` which was historically always installed with Linux distributions. FSL compiles without this, but Sienna will fail without it.

**Comment by neurolabusc** [comment]
Nice find. Does changing the makefile so it includes both the x86 and ARM paths fix this? I am busy today, but can test tomorrow. Would be nice to have `vastness` supported.

So my suggestion is:

```
SUMA_INCLUDE_PATH = -I/usr/X11R6/include -I./ -I../ -I../niml/ -Igts/src -I/usr/include/glib-1.2 -I/usr/include/glib-2.0 -I/usr/lib64/glib/include -I/usr/lib64/ glib-2.0/include -I/usr/lib/x86_64-linux-gnu/glib-2.0/include/ -I/usr/lib/aarch64-linux-gnu/glib-2.0/include
```

**Comment by afni-rickr** [comment]
Makefile.INCLUDE is not fully protected against a partial success.
Delete libjpeg.a to remake cjpeg and djpeg.
What system are you building on?  If Fedora or CentOS, consider the prereq files other_builds/OS_notes.linux_fedora_35.txt or other_builds/OS_notes.linux_centos_7.txt.

**Comment by afni-rickr** [comment]
No, I will fix that in the Makefile, so it will not have to be a build concern.  Right now it assumes that if libjpeg.a exists, then so does cjpeg and djpeg, so it will not "remake" them.
If you would like, I will check in that update and you can merge master into whatever you are doing.  Would that work?

**Comment by afni-rickr** [comment]
I did push that update to master.  It has been a minor irritation over the years...

**Comment by afni-rickr** [comment]
Thanks @markjens.  I will let @neurolabusc comment on the Makefile updates, until I get systems for testing.
The vararg_ptr upated lines are good, though I will actually just remove them.  It seems like that was a mistaken assumption of vararg_ptr actually being a pointer (but it is a struct), which currently always evaluates to true.
For getcwd(), it could return NULL if the function returns that.  Alternatively, it could reallocate until it has space, but feels uncomfortable, too.
Would you like to make changes with a pull request, or would you prefer that I make them?

**Comment by afni-rickr** [comment]
@neurolabusc Makefile.INCLUDE was updated yesterday to include [cd]jpeg in the target.  See [https://github.com/afni/afni/commit/025ce4ad3c4d4f8ba064c8e9eddc573b51427474](https://github.com/afni/afni/commit/025ce4ad3c4d4f8ba064c8e9eddc573b51427474).

**Comment by martin-g** [comment]
Hi!

I was able to build AFNI on openEuler 20.03 LTS ARM64 with small modifications in `other_builds/Makefile.linux_ubuntu_22_64`:

```diff
diff --git src/other_builds/Makefile.linux_ubuntu_22_64 src/other_builds/Makefile.linux_ubuntu_22_64
index 706665366..bfcbb3307 100644
--- src/other_builds/Makefile.linux_ubuntu_22_64
+++ src/other_builds/Makefile.linux_ubuntu_22_64
@@ -27,7 +29,7 @@ R_IO_MODIFY_LINUX = yes
 # L_USE_ACML = /opt/acml3.1.0/gnu64/lib/libacml.a
 
 CCDEBS = -DAFNI_DEBUG -DIMSEQ_DEBUG -DDISPLAY_DEBUG -DTHD_DEBUG
-CEXTRA = -Wcomment -Wformat -DUSE_TRACING -DHAVE_XDBE -D_GNU_SOURCE \
+CEXTRA = -Wcomment -Wformat -DUSE_TRACING -D_GNU_SOURCE \
          -DREPLACE_XT $(CPROF)
 
 #use gcc by default unless user demands clang or gcc not found
@@ -131,12 +133,12 @@ LLIBS_X11 = -lmri -lf2c -lXm -lXpm -lXext -lXmu -lXt -lSM -lICE -lX11 \
 # vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv
 # For suma (NO STATIC LINKING OF GL libs)
 SUMA_GLIB_VER = -2.0
-SUMA_INCLUDE_PATH = -I/usr/X11R6/include -I./ -I../ -I../niml/ -Igts/src -I/usr/include/glib-1.2 -I/usr/include/glib-2.0 -I/usr/lib64/glib/include -I/usr/lib64/ glib-2.0/include -I/usr/lib/x86_64-linux-gnu/glib-2.0/include/
+SUMA_INCLUDE_PATH = -I/usr/X11R6/include -I./ -I../ -I../niml/ -Igts/src -I/usr/include/glib-1.2 -I/usr/include/glib-2.0 -I/usr/lib64/glib/include -I/usr/lib64/ -I/usr/lib/x86_64-linux-gnu/glib-2.0/include/  -I/usr/lib/aarch64-linux-gnu/glib-2.0/include/
 SUMA_LINK_PATH = -L/usr/lib64 -L/usr/X11R6/lib64 -L../
 #use -lGLw if you have libGLw.a or libGLw.so* or 
 #  -lMesaGLw if you have Mesa's version (libMesaGLw*) of libGLw
 #
```

I had to remove `-DHAVE_XDBE` because of the following error:
```
/bin/ld: libXmHTML.a(LZWStream.o): in function `LZWStreamInit':
LZWStream.c:(.text+0xbc8): warning: the use of `tmpnam' is dangerous, better use `mkstemp'
/bin/ld: ./libmrix.a(xim.o): in function `memplot_to_X11_funfunfun':
xim.c:(.text+0x7604): undefined reference to `getwin_from_XDBE'
collect2: error: ld returned 1 exit status
make: *** [Makefile.INCLUDE:1853: afni] Error 1
```
and I wasn't able to find which dependency is supposed to provide this function.

The next issue was that openEuler does not provide `mesa-libGLw-devel` anymore...
https://gitee.com/src-openeuler/mesa-libGLw seems to be abandoned since few years.
According to https://gitee.com/openeuler/community/pulls/1411 the reason was that the upstream project is not maintained any more,
I worked it around by installing the .rpm from http://mirror.centos.org/altarch/7/os/aarch64/Packages/mesa-libGLw-devel-8.0.0-5.el7.aarch64.rpm

**Comment by martin-g** [comment]
> until I get systems for testing.

@afni-rickr Please let me know if you are interested in my offer at https://github.com/afni/afni/issues/390#issuecomment-1212928019 !

**Comment by afni-rickr** [comment]
Thanks @martin-g, I will let you know!

**Comment by afni-rickr** [comment]
@markjens, those changes are here: [](https://github.com/afni/afni/commit/7ab354956bf89ff3174b027bde28efebc344ca30)https://github.com/afni/afni/commit/7ab354956bf89ff3174b027bde28efebc344ca30.  Thanks!

**Comment by neurolabusc** [comment]
@markjens I was able to get your Makefile to work for totality on Ubuntu 22.04 for both arm64 and x86-64 with the following caveats. Maybe @afni-rickr has some ideas to resolve these.
 1. Needed libgtk2.0-dev - be aware that GTK2 is deprecated and may behave oddly and be aware of fix [for slow startup times](https://askubuntu.com/questions/1184774/some-applications-on-ubuntu-19-10-very-slow-to-start).
 2.  I had `multiple definition of glwMDrawingAreaWidgetClass errors` (for both architectures) which I fixed with this [hack of /usr/include/GL/gl.h](https://stackoverflow.com/questions/36562111/multiple-definition-of-glwmdrawingareawidgetclass). My recipe clearly creates some sort of library conflict that I can not resolve.
 3. @martin-g we are able to complied with `-DHAVE_XDBE`. I did face this issue earlier, and one of the libraries in my recipe fixed it, but I can not remember which one.

```
git clone https://github.com/afni/afni.git
cd afni/src
cp Makefile.linux_ubuntu_22 Makefile
sudo apt-get update
sudo apt-get install -y tcsh xfonts-base libssl-dev       \
                        python3-matplotlib python3-numpy  \
                        gsl-bin netpbm   \
                        libjpeg62 xvfb xterm vim curl     \
                        gedit evince eog                  \
                        libglu1-mesa-dev libglw1-mesa     \
                        libxm4 build-essential            \
                        libcurl4-openssl-dev libxml2-dev  \
                        gnome-terminal nautilus libgomp1  \
                        gnome-icon-theme-symbolic         \
                        firefox xfonts-100dpi libxtst-dev \
                        r-base-dev cmake libxt-dev        \
                        libgdal-dev libopenblas-dev       \
                        libudunits2-dev       \
                        libxext-dev libmotif-dev          \
                        libxpm-dev libxmu-dev libz-dev    \
                        libxmu-headers libgsl-dev make    \
                        libexpat1-dev libmotif-dev
#for Ubuntu 22.04
sudo apt-get install -y libgtk2.0-dev python-is-python3 libnode-dev 
#for gcc build
make all
#for clang build
sudo apt-get install -y clang libomp-dev libomp-dev
useClang=1 make all
```

**Comment by afni-rickr** [comment]
Hi @neurolabusc.  That glwMDrawingAreaWidgetClass issue has been around for a few years, and is referenced in some of our install instructions, such as [OS_notes.linux_ubuntu_16_64](https://github.com/afni/afni/blob/master/src/other_builds/OS_notes.linux_ubuntu_16_64).  It seems to have been resolved on some of the more recent systems, such as Fedora 35.

**Comment by neurolabusc** [comment]
@afni-rickr I missed that. I have updated my recipe. The updated makefile looks like it works for both gcc/clang, x86-64/arm64 and Ubuntu 18.04-22.04. I think this issue generated a lot of great teamwork that should help a lot of users.

**Comment by martin-g** [comment]
> 3\. we are able to complied with `-DHAVE_XDBE`. I did face this issue earlier, and one of the libraries in my recipe fixed it, but I can not remember which one.

@neurolabusc  I have this problem on openEuler OS, which is RHEL based.
I have tried your recipe at https://github.com/afni/afni/issues/390#issuecomment-1207239922 and the official ones for [RHEL](https://afni.nimh.nih.gov/pub/dist/doc/htmldoc/background_install/install_instructs/steps_linux_RH.html#install-prerequisite-packages) and [Fedora](https://afni.nimh.nih.gov/pub/dist/doc/htmldoc/background_install/install_instructs/steps_linux_Fed.html#slow-setup-install-prerequisite-packages). I will try again next week.

**Comment by martin-g** [comment]
It appears that one of the fixes for the Makefile was never committed - #421 !
Indeed having a CI for this `universal` Makefile would have caught this! But I understand that having CI for all available Makefiles would be too much! 
From one side I think this universal Makefile deserves to be tested by CI because it covers Ubuntu 18.04-22.04+ and modern RedHat-like distros (like Fedora, Rocky, openEuler, Alma, ...) but on the other side since no one faced this issue until now then maybe it is not that important...

**Comment by adityaroy10** [comment]
The Makefile work here already landed (`Makefile.linux_ubuntu_24_ARM`), thanks for that. What's still missing is that the official `afni_dev_base`/`afni_make_build` Docker images are never built for arm64, only amd64, so `docker pull` still gives you an emulated image on Apple Silicon or ARM Linux even though the source builds natively.

I'm working on this as part of a broader effort to get fMRIPrep running natively on `linux/arm64` (it currently only ships `linux/amd64` and runs under emulation on Apple Silicon, confirmed by several long-open user reports on that repo). fMRIPrep pulls AFNI from `afni_make_build`, so this is the piece blocking it there, but a real arm64 image benefits anyone running AFNI directly on ARM, not just fMRIPrep.

I checked the Dockerfile and CI: no arm64-specific package gaps in `afni_dev_base` (verified the apt list against Ubuntu 18.04's arm64 archive), one hardcoded x86_64 CMake download that needs a conditional swap, and two unrelated package-name fixes (`git-annex-standalone` isn't on arm64, use `git-annex`; `ncurses-dev` isn't a real package name on any architecture, should be `libncurses-dev`). The bigger piece is `.circleci/config.yml`, which only builds/pushes a single amd64 image today, no arm64 job at all. Planning to open a PR adding that. Flagging first in case anyone's already on it.
